Add completed tasks

chuazhan opened the pull request:
Add completed tasks to check if preceding tasks are completed or not.
There is an old logic in workflow.py:
if len(join_task.preceded_by) == len(join_list):
It is used to check if current task is blocked or not, but there is an exceptional case when removing a subtask before a join point, original len(join_task.preceded_by)-len(join_list)=1, after removing a preceding task, the two length will be equal which result in wrong logic.
This patch add a completed tasks and check if preceding tasks names are in completed tasks instead of checking the length.

Change the tests' task names in test_waiting_branches.py as context.done will be affected by tests in test_waiting.py.

Some minor flake8 issue as well.

JIRA: RHELWF-7410

hluk commented on tests/test_runner/test_waiting_branches.py:
```diff
@@ -68,14 +68,17 @@ def test_run(self):
+        self.assertSetEqual(context.done, {'start_wb',
+                                           'blocked_wb',
+                                           'not_blocked_wb',
+                                           'end_wb'})
```
Was there actually any functional change in this file? Why were all the tasks renamed?

chuazhan replied:
I answered in another RP, paste it here:
This is weird. That is why I pushed another branch in RHELWF-7410_add_completed_tasks_set, I'd like to show you the issue. You can test it locally and see the test results.
It seems test_waiting_branches.py will be affected by tests in test_waiting.py.
context.done will be {'task_bb', 'task_cd', 'blocked', 'task_a', 'task_d', 'first', 'not_blocked', 'fail', 'end', 'task_x', 'task_c', 'task_b', 'task_ab', 'start', 'second'} in self.assertSetEqual(context.done, {'start', 'blocked', 'not_blocked'})
But actually task like 'task_bb' are not in test_waiting_branches.py, I am not sure why it happens, maybe related to decorator, I am not sure.
Rename the task names or comment out test 'test_run' in test_waiting.py will fix the issue.

hluk replied:
When does this happen? For me running `tox` passes all tests. Is this related to your change?

chuazhan replied:
Yes. After adding my change, this happens, I haven't figured out the reason :(

hluk commented on wfepy/workflow.py:
```diff
@@ -125,13 +127,15 @@ def load(self, file_path):
 
     def dump(self, file_path):
```
Do we actually use this to store the execution state?

I see only `state.json` in our Jira issues but it does not contain `context` only `state` (and some customized attributes).

chuazhan replied:
No. I think our codes override it. It contains state, subtask, data, etc. I am thinking to change our codes as well.

hluk commented:
Can we keep this backwards-compatible and migrate existing in-progress states? Maybe also adding `wfepy_version` attribute would help.

chuazhan commented:
> Can we keep this backwards-compatible and migrate existing in-progress states? Maybe also adding `wfepy_version` attribute would help.

I don't understand. Do you mean we don't change it in wpefy?

hluk commented on wfepy/workflow.py:
```diff
@@ -108,11 +108,13 @@ class Runner:
     workflow = attr.ib()
     context = attr.ib(default=None)
     state = attr.ib(default=None, init=False)
+    completed_tasks = attr.ib(default=[])
```
The problem with tests is that the default list is actually used by instances and changes persist for later instances (i.e. the list is not copied). Fix:

```suggestion
    completed_tasks = attr.ib(factory=list)
```

chuazhan replied:
Ah. Ok, Thanks @hluk

chuazhan committed:
Add completed tasks

Add completed tasks to check if preceding tasks are completed
or not. (Note: completed tasks include completed/canceled
non-joint tasks.)
There is an old logic in workflow.py:
if len(join_task.preceded_by) == len(join_list):
It is used to check if current task is blocked or not, but there
is an exceptional case when removing a subtask before a join
point, original len(join_task.preceded_by)-len(join_list)=1,
after removing a preceding task, the two length will be equal
which result in wrong logic.
This patch add a completed tasks and check if preceding tasks
names are in completed tasks instead of checking the length.

Some minor flake8 issue as well.

JIRA: RHELWF-7410

hluk commented:
> > Can we keep this backwards-compatible and migrate existing in-progress states? Maybe also adding `wfepy_version` attribute would help.
> 
> I don't understand. Do you mean we don't change it in wpefy?

No, better to fix this here. But what about the tasks executed with previous wfepy version we want to resume? These do not contain the `completed_tasks` in stored state.

chuazhan commented:
> > > Can we keep this backwards-compatible and migrate existing in-progress states? Maybe also adding `wfepy_version` attribute would help.
> > 
> > 
> > I don't understand. Do you mean we don't change it in wpefy?
> 
> No, better to fix this here. But what about the tasks executed with previous wfepy version we want to resume? These do not contain the `completed_tasks` in stored state.

Ah, Ok. Good point, I got it. Maybe we can construct them according to the state.json file.

hluk commented on wfepy/workflow.py:
```diff
@@ -242,6 +246,7 @@ def _step(self, state):
                 elif result:
                     logger.info('Task %s is complete', task_name)
                     next_state.append((task_name, TaskState.COMPLETE))
+                    self.completed_tasks.append(task_name)
```
Modifying `self.completed_tasks` in this method does not seem correct because it will be out-of-sync with `self.state` potentially.

hluk replied:
Would it be possible to keep COMPLETE in state instead and later see if we can add NEW tasks based on this?

chuazhan replied:
Close this patch and think about to re-implement the whole looping logic
